### ISSUE TITLE
XML Decompiler

### ISSUE THREAD

**Issue by biden2020prez** [MuntashirAkon/AppManager#1045]
### Please check before submitting an issue

- [X] I am using the latest version of App Manager
- [X] I have searched the issues and haven't found anything relevant
- [X] I have read the docs

### Describe a description of the new feature

Not sure if this was included in the plan or not, but - It would be really nice if we can open an xml file from file manager "open with", and choose app manager. Then it decompiles that binary xml file, we edit it, and it recompiles. Let me know what you think

To be clear, I found another relevant issue, but the developer of this project requested opening a new feature for all issues, because there was too much to keep track of.

**Comment by MuntashirAkon** [comment]
Yes, this is very much possible with the newly added code editing feature.

**Comment by biden2020prez** [comment]
> Yes, this is very much possible with the newly added code editing feature.

amazing. it will be very useful to edit standalone xml files without decompiling a entire apk. Thank you so much!!

**Comment by MuntashirAkon** [comment]
> it will be very useful to edit standalone xml files without decompiling a entire apk.

Yes, we're now very close to offer APK editing (#138) within App Manager.

**Comment by biden2020prez** [comment]
Amazing, thanks for all your hard work on this app.

**Comment by Gabs-Unk** [comment]
must agree with that, so uselful for me and a lot of users I am pretty sure !

**Comment by MuntashirAkon** [comment]
This feature is now available in the latest alpha (v4.0.0-alpha01)

**Comment by biden2020prez** [comment]
Hey, thanks for this feature. On the samsung galaxy s9 where i tried this, I couldnt save any edited files to the location i opened them. it would show an error dialog telling me app manager doesnt have write access and prompting me to choose another location. 

And when i navigated to the location that the file is, it doesn't overwrite, instead it adds (1) to the filename, without prompting anything. 


should this be a new issue?

thanks for the amazing app and amazing new feature by the way

**Comment by biden2020prez** [comment]
I will add (in a new comment just in case you already looked at the previous one) that my phone is running android 9.

The XML was inside of /sdcard/android/data folder, which shouldn't be a problem in android 9. But to make sure, I tested xmls in the /sdcard and /sdcard/Download folder and they both had the same issue. 

this issue was with binary xmls as well as normal .txt files. Additionally, I had this issue whether I opened App Manager's code editor from another file manager or whether I opened it from App Manager's file manager.

I had my friend test it, and he also couldn't overwrite an existing file in his Android 9 and Android 11 phone. 

I am afraid I just didnt set something correctly in App Manager, but I couldn't find a setting that fixes it. 

App Manager works fine to save to a new file, just not an exisiting one. Storage access is granted in settings. 

thank you so much @MuntashirAkon and if you need any more information don't hesitate to ask.

**Comment by dds1236** [comment]
@MuntashirAkon l am having the same error, here is a screenshot
![Screenshot_20230721_112539](https://github.com/MuntashirAkon/AppManager/assets/101875008/d8a4863d-76bc-4ff7-8fea-eab824fb2b1b)
I'm using Android 11

**Comment by MuntashirAkon** [comment]
This seems to be happening only in the release builds. This is probably why I haven't noticed it because it's not producible in the debug builds. For some reason, in the release builds, `Path#canWrite()` is always returning `false`.

**Comment by MuntashirAkon** [comment]
> Additionally, I had this issue whether I opened App Manager's code editor from another file manager or whether I opened it from App Manager's file manager.

Seems to be working when a file is opened from the file managers supply files using a content provider (Material Files, MiXplorer, etc.)

**Comment by biden2020prez** [comment]
> in the release builds, `Path#canWrite()` is always returning `false`.

I googled this, it may be a Java bug 
https://stackoverflow.com/questions/60849218/canwrite-not-returning-value-as-expected


Although other stackoveflow pages suggested other things. the main thing I saw in a few places was that  Files.isWritable() is more reliable than File.canWrite(). 

I know you know way more about java and android than I do, just trying to be helpful.



> This seems to be happening only in the release builds. This is probably why I haven't noticed it because it's not producible in the debug builds.

I guess I can try building a debug build and see what happens.



> Seems to be working when a file is opened from the file managers supply files using a content provider (Material Files, MiXplorer, etc.)

I'll try one of these later



Thanks so much for your hard work!

**Comment by MuntashirAkon** [comment]
> I googled this, it may be a Java bug https://stackoverflow.com/questions/60849218/canwrite-not-returning-value-as-expected
> 
> Although other stackoveflow pages suggested other things. the main thing I saw in a few places was that Files.isWritable() is more reliable than File.canWrite().

`Path` is a custom in-house API, not related to those Java APIs. Anyway, I'm currently taking a break. I'll fix it in the next month.

**Comment by biden2020prez** [comment]
Okay, I just made a debug build which didn't immediately solve the issue. I'll look at the code and see if I can figure it out myself, because this feature is really exciting!


I hope you are taking a _good_ break, I hope you enjoy it. All the best!

**Comment by MuntashirAkon** [comment]
Does not seem to reproduce it anymore or I've forgotten the steps to reproduce the issue.

**Comment by biden2020prez** [comment]
> Does not seem to reproduce it anymore or I've forgotten the steps to reproduce the issue.

I had the issue simply downloading the v4.0.0-alpha01 version, and using it. Also had the issue when compiling myself. In the end, since im bad at java,  I made it always return true that it can write, which was a bad solution, but temporarily allowed me to use the feature while I waited for a fix.

when you use [you pre-release apk](https://github.com/MuntashirAkon/AppManager/releases/download/v4.0.0-alpha01/AppManager_v4.0.0-alpha01.apk), and from a file manager press open with, then code editor (from appmanager), then edit something and press save, its saving properly?

Maybe you are having trouble reproducing it because you are trying with one of these file managers (i meant to try but i didnt yet)?




> Seems to be working when a file is opened from the file managers supply files using a content provider (Material Files, MiXplorer, etc.)


Thanks for all your amazing work!!!
